src/CollateralFactory.sol

```diff
@@ -24,60 +48,123 @@ abstract contract Setter {
+    Setter public immutable taxCollector;
+    Setter public immutable liquidationEngine;
+    Setter public immutable oracleRelayer;
+    Setter public immutable globalSettlement;
+    Setter public immutable stabilityFeeTreasury;
+    FactoryLike public immutable osmFactory;
+    FactoryLike public immutable joinFactory;
+    FactoryLike public immutable auctionHouseFactory;
+    FactoryLike public immutable keeperIncentivesFactory;
+
+    event contractDeployed(string, address);
 
     constructor(
+        address pauseProxy_,
         address safeEngine_,
         address liquidationEngine_,
         address oracleRelayer_,
         address globalSettlement_,
         address taxCollector_,
+        address stabilityFeeTreasury_,
+        address osmFactory_,
         address joinFactory_,
-        address auctionHouseFactory_
+        address auctionHouseFactory_,
+        address keeperIncentivesFactory_
     ) public {
+        pauseProxy = pauseProxy_;
         safeEngine = Setter(safeEngine_);
         liquidationEngine = Setter(liquidationEngine_);
         oracleRelayer = Setter(oracleRelayer_);
         globalSettlement = Setter(globalSettlement_);
         taxCollector = Setter(taxCollector_);
+        stabilityFeeTreasury = Setter(stabilityFeeTreasury_);
+        osmFactory = FactoryLike(osmFactory_);
         joinFactory = FactoryLike(joinFactory_);
         auctionHouseFactory = FactoryLike(auctionHouseFactory_);
+        keeperIncentivesFactory = FactoryLike(keeperIncentivesFactory_);
     }
 
-    function deployCollateralType(
-        address token,
-        address osm,
+    // notice: this function can be called by anyone, it will deploy a set of contracts needed for a new collateral (no impact on the system until they are attached)
+    function deployCollateralSpecificContracts(
         bytes32 collateralType,
+        address token,
+        address priceFeed,
+        address coinOracle,
+        address ethOracle
+    )
+        external
+        returns (
+            address join,
+            address auctionHouse,
+            address osm,
+            address keeperIncentives
+        )
+    {
+        join = joinFactory.deploy(
+            address(safeEngine),
+            collateralType,
+            token,
+            pauseProxy
+        );
+        emit contractDeployed("JOIN", join);
+
+        auctionHouse = auctionHouseFactory.deploy(
+            address(safeEngine),
+            address(liquidationEngine),
+            collateralType,
+            pauseProxy
+        );
+        emit contractDeployed("AUCTION_HOUSE", auctionHouse);
+
+        osm = osmFactory.deploy(priceFeed, pauseProxy);
+        emit contractDeployed("OSM", osm);
+
+        keeperIncentives = keeperIncentivesFactory.deploy(
+            address(stabilityFeeTreasury),
+            osm,
+            address(oracleRelayer),
+            [collateralType, bytes32(0), bytes32(0)],
+            2 * 10 ** 18,
```

Comment:
I think this should be increased slightly to `4 * 10 ** 18`
Since we aren't emitting ERC-20 rewards, the caller must still pay `CoinJoin.exit()` costs. We also aren't accounting for any tip the caller must use, and are just using the basefee. Finally, we are using the TWAP as the `coinOracle` to calculate rewards and there could be divergence from current market prices.

Reply:
Yes, when I start deploying the contracts I will deploy a 5m twap for us to use as good approximation to the market price. We used 2 wad for the other incentives, are you sure we should deviate on this one? Re' calling exit, I think they will either exit on the same tx or accumulate before calling it, this should alleviate the cost.